### PR TITLE
Ensure sendBatch not called recursively

### PR DESCRIPTION
**Pull request by dnhatn**
This PR introduces AsyncRecoveryTarget which executes remote calls of peer recovery asynchronously. In this change, we also add a new assertion to ensure that method `sendBatch`, which sends a batch of history operations in phase2, is never called recursively on the same thread. This new assertion will also be used in method `sendFileChunks`.

Relates https://github.com/elastic/elasticsearch/pull/39769#discussion_r264150661

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -584,6 +584,7 @@ private void sendBatch(
             final long maxSeqNoOfUpdatesOrDeletes,
             final RetentionLeases retentionLeases,
             final ActionListener<Long> listener) throws IOException {
+        ThreadPool.assertCurrentMethodIsNotCalledRecursively();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: this should be ` assert ThreadPool.assertCurrentMethodIsNotCalledRecursively();`

**Review by s1monw — APPROVED** [review]
Left one comment

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Yes, fixed in https://github.com/elastic/elasticsearch/pull/39988/commits/6109f0ee68ac8fe7760bfcf8206d95964cbf9c87.

**Comment by dnhatn** [comment]
Thanks @s1monw.

**Pull request merged**